### PR TITLE
VZ-4747 - Add an affinity configuration for the AuthProxy deployment

### PR DESCRIPTION
**Pull request by mgianatagh**
# Description

Add an affinity configuration for the AuthProxy deployment.  The option to customize the affinity configuration via The Verrazzano custom resource will be handled in a separate task.

# Checklist 

As the author of this PR, I have:

- [ ] Checked that I included or updated copyright and license notices in all files that I altered
- [ ] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: platform-operator/helm_config/charts/verrazzano/values.yaml [hunk 1]
```diff
@@ -105,6 +105,18 @@ api:
     AuthnStateTTL: "300"
     MaxRequestSize: 65m
     ProxyBufferSize: 8k
+  affinity: |
+    podAntiAffinity:
+      preferredDuringSchedulingIgnoredDuringExecution:
```
Comment (theabmitra) [inline platform-operator/helm_config/charts/verrazzano/values.yaml:110]: A scenario to consider here .
preferredDuringSchedulingIgnoredDuringExecution is a soft check by the scheduler, so in a scenario if number of nodes < number of pods, the pods will be assigned successfully on that node . 
However they will remain like that unless some external action is taken . 
Just like we are adding checks in go-code to check if 2 replicas are running with affinity configured, we may also have this check and just scale down -> scale up the deployment only when number of nodes > number of pods.

File: tests/e2e/pkg/kubernetes.go [hunk 2]
```diff
@@ -128,6 +128,36 @@ func ListDeployments(namespace string) (*appsv1.DeploymentList, error) {
...
+		return nil, err
+	}
+	return deployment, nil
+}
+
+// DoesDeploymentExist returns whether a deployment with the given name and namespace exists for the cluster
+func DoesDeploymentExist(namespace string, name string) (bool, error) {
+	deployments, err := ListDeployments(namespace)
+	if err != nil {
+		Log(Error, fmt.Sprintf("Error listing deployments in cluster for namespace: %s, error: %v", namespace, err))
```
Comment (pfmackin) [inline tests/e2e/pkg/kubernetes.go:150]: should be "Failed listing.  You don't need error: %v,  just : %v

File: tests/e2e/pkg/kubernetes.go [hunk 3]
```diff
@@ -128,6 +128,36 @@ func ListDeployments(namespace string) (*appsv1.DeploymentList, error) {
...
+// GetDeployment returns a deployment with the given name and namespace
+func GetDeployment(namespace string, deploymentName string) (*appsv1.Deployment, error) {
+	// Get the Kubernetes clientset
+	clientSet, err := k8sutil.GetKubernetesClientset()
+	if err != nil {
+		return nil, err
+	}
+	deployment, err := clientSet.AppsV1().Deployments(namespace).Get(context.TODO(), deploymentName, metav1.GetOptions{})
+	if err != nil {
+		Log(Error, fmt.Sprintf("Failed to get Deployment %v from namespace %v:  Error = %v ", deploymentName, namespace, err))
```
Comment (pfmackin) [inline tests/e2e/pkg/kubernetes.go:140]: don't need : Error = %v,  just : %v

**Review by pfmackin — APPROVED** [review]

File: tests/e2e/pkg/kubernetes.go (hunk 3 above)
Comment (mgianatagh) [reply tests/e2e/pkg/kubernetes.go:140]: fixed in my next push.  I am reviewing all the messages in the file to fix them according to your feedback.

File: platform-operator/helm_config/charts/verrazzano/values.yaml (hunk 1 above)
Comment (kminder) [reply platform-operator/helm_config/charts/verrazzano/values.yaml:110]: This goes back to the discussion of preferredDuringSchedulingIgnoredDuringExecution vs requiredDuringSchedulingIgnoredDuringExecution.  Ideally I think a prod profile should use requiredDuringSchedulingIgnoredDuringExecution but this won't work if someone tries to use prod on a single node cluster (even though they shouldn't).  In the end I think this is something that we should expect customer to tweak and make it easy for them.  Specifically I imagine most "real" customer will switch to requiredDuringSchedulingIgnoredDuringExecution and add zone anti-affinity (i.e. topology.kubernetes.io/zone).

**Review by pfmackin — DISMISSED** [review]

**Review by kminder — APPROVED** [review]

File: tests/e2e/pkg/kubernetes.go [hunk 4]
```diff
@@ -268,7 +298,7 @@ func GetVerrazzanoInstallResourceInCluster(kubeconfigPath string) (*v1alpha1.Ver
 func IsDevProfile() bool {
 	kubeconfigPath, err := k8sutil.GetKubeConfigLocation()
 	if err != nil {
-		Log(Error, fmt.Sprintf("Error getting kubeconfig, error: %v", err))
+		Log(Error, fmt.Sprintf("Error getting kubeconfig: %v", err))
```
Comment (kminder) [inline tests/e2e/pkg/kubernetes.go:301]: Should this start with "Failed" too?

File: tests/e2e/pkg/kubernetes.go [hunk 5]
```diff
@@ -286,7 +316,7 @@ func IsDevProfile() bool {
 func GetVerrazzanoVersion() (string, error) {
```
Comment (kminder) [inline tests/e2e/pkg/kubernetes.go:316]: Nothing to do with your change bu

File: tests/e2e/pkg/kubernetes.go [hunk 6]
```diff
@@ -324,7 +354,7 @@ func IsVerrazzanoMinVersion(minVersion string) (bool, error) {
 func IsProdProfile() bool {
```
Comment (kminder) [inline tests/e2e/pkg/kubernetes.go:354]: Nothing to do with your change but these should probably return the error instead of defaulting to false.

**Review by kminder — APPROVED** [review]
Approved as fixing the logging is really out of scope of the story.

**Pull request merged**